File: Top Interview Questions/33. Search in Rotated Sorted Array/solution.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int low=0,high=nums.size()-1;
        while(low < high) {
            int mid = (high - low) / 2 + low;
            if(nums[mid] > nums[high]) low = mid + 1;
            else high = mid;
        }
        int rot = low;
        low = 0, high = nums.size()-1;
        while(low <= high) {
            cout << low << ' ' << high << endl;
            int temp_mid = (high - low) / 2 + low;
            int mid = (temp_mid + rot) % nums.size();
            if(nums[mid] == target) return mid;
            else if(nums[mid] < target) low = temp_mid + 1;
            else high = temp_mid - 1;
        }
        return -1;
    }
};

File: Top Interview Questions/33. Search in Rotated Sorted Array/solution.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:

        low,high = 0, len(nums) - 1
        
        while low < high:
            mid = low + (high - low) // 2
            if nums[mid] < nums[high]:
                high = mid
            else:
                low = mid + 1
        
        rotated = low

        low, high = 0, len(nums) - 1
        while low <= high:
            mid = low + (high - low) // 2
            rotated_mid = (mid + rotated) % len(nums)
            
            if nums[rotated_mid] == target: return rotated_mid
            elif nums[rotated_mid] < target: low = mid + 1
            else: high = mid - 1

        return -1



s = Solution()

nums = [4,5,6,7,0,1,2]
target = 0
print(s.search(nums,target))
